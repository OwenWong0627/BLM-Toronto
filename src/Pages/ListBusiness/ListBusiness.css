.title{
   color: #000;
   font-family: 'Righteous', cursive;
   font-size: 65px; font-weight: normal;
   line-height: 60px;
   text-transform: uppercase;
   text-shadow: 2px 2px 0 grey;
   text-align: center;
   margin: 7.5vh 2.5vw;
}

.selectTitle{
   color : #082567;
   font-family: 'Righteous', cursive;
   font-size: 32px; font-weight: normal;
   margin: 0 0 10px;
   text-transform: uppercase;
   text-shadow: 1px 1px 0 grey;
   text-align: center;
}

.container ul{
  list-style-type: none;
}

.BusinessList {
  padding: 0; /* Remove padding */
  margin: 0 4rem; /* Remove margins */
}

.Business {
  border: 1px solid #ddd; /* Add a thin border to each list item */
  margin-top: -1px; /* Prevent double borders */
  background-color: #f6f6f6; /* Add a grey background color */
  padding: 12px; /* Add some padding */
}

.Business a {
   text-decoration: none;
}

.Business button {
   border: none;
   background-color: Transparent;
}

.title-icon {
   display: flex;
   justify-content: space-between;
   align-items: center;
}

.Business img {
   height: 50px;
   margin-bottom: -20px;
   padding: 5.75px 4.5px 5px 5px;
}

.Business img:hover {
   background-color: lightgray;
   transition: background-color .3s;
   border-radius: 100%;
   cursor: pointer;
}

.pagination {
  width: 100vw;
  text-align: center;
  margin: 7.5vh 0;
}

.pagination li{
  display: inline-block;
}

.pagination a {
  color: black;
  float: left;
  padding: 8px 16px;
  text-decoration: none;
  transition: background-color .3s;
  border: 1px solid #ddd;
  margin: 0 4px;
  text-align: end;
}

.pagination a:hover {
  background-color: #00BFFF;
  color: white;
  border: 1px solid #00BFFF;
}

.pagination a.activePage{
   pointer-events: none;
   background-color: #00BFFF;
   color: #fff;
   border: 1px solid #00BFFF;
}

/* Only a temporary solution as pointer event only "disables" you being able to click on the element,
but it can still be accessed by other methods, like using keyboard */
.button.inactive {
   pointer-events: none;
   color: #aaafb4;
}

select {
   -webkit-appearance: none;
   -moz-appearance: none;
   -ms-appearance: none;
   appearance: none;
   outline: 0;
   box-shadow: none;
   border: 0;
   background: #CDCDCD;
   background-image: none;
}
/* Remove IE arrow */
select::-ms-expand {
  display: none;
}
/* Custom Select */
.select {
  position: relative;
  width: 50%;
  height: 3em;
  line-height: 3;
  background: #CDCDCD;
  overflow: hidden;
  border-radius: .5em;
  display: flex;
  margin: 0 auto 40px auto;
}
.select-text{
   font-size: 25px;
}
select {
  flex: 1;
  padding: 0 .5em;
  color: #000;
  size: 100px;
  cursor: pointer;
}

select > option{
   font-size: 20px;
}
select > text{
   font-size: 20px;
}

/* Arrow */
.select::after {
  content: '\25BC';
  position: absolute;
  top: 0;
  right: 0;
  padding: 0 1em;
  background: #00bfff;
  color: white;
  cursor: pointer;
  pointer-events: none;
  -webkit-transition: .4s all ease;
  -o-transition: .4s all ease;
  transition: .4s all ease;
}
/* Transition */
.select:hover::after {
  color: #000;
}