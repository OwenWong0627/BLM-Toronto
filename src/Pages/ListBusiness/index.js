//TODO: Currently, we are doing front-end pagination which will call all of the data and display
//[number] of businesses at a time. In the future, we would make it scalable for a lot of businesses
//in a database by calling the database and displaying the called [number] of businesses.
import React, { useState, useEffect } from 'react';
import * as businessData from '../../businesses.json';
import Business from '../../Components/Business';
import './ListBusiness.css';
import ScrollToTopOnMount from '../../Components/ScrollToTopOnMount';

const IPAPIURL = `https://ipinfo.io/json?token=${process.env.REACT_APP_IPINFO_KEY}`;

const allCities = businessData.allCities;

/**
 * This is function that calls on an IPAPI to extract the latitude and longitude data from the current user
 * This api will avoid prompting the user to allow the "This app wants to know your location" function
 * However, it is important to note that IPAPI do not give completely accurate locations
 *
 * @async
 * @param {string} URL - The URL to receive the user's location details from
 * @return {Promise.<Object>} - The latitude and longitude data from the URL
 */
async function getUserLocation(URL) {
   let response = await fetch(URL);
   let data = await response.json();
   console.log(data);
   return ({
      lat: parseFloat(data.loc.split(",")[0]),
      lng: parseFloat(data.loc.split(",")[1])
   });
}

/**
 * This function calculates the nearest city to a location, origin, by searching the cities array and finding the city that has the least distance from the origin
 * 
 * @param {Object.<string, number>} origin - An object containing lat and lng keys that has float values
 * @param {Array.<Object>} cities - An array of city objects that has key for the city name, and a key for city's lat and lng values
 * @returns {string} - The function returns a string with the nearest city from the user's location
 */
export function getNearestCity(origin, cities) {
   //Linear Search but not really, because it will execute with a complexity of O(n) to find the nearest city
   for (let i = 0; i < cities.length; i++) {
      if (typeof (cities[i].LatLng.lat) !== "number" || typeof (cities[i].LatLng.lng) !== "number") {
         console.log('The lat/lng must be numbers');
         return null;
      }
   }
   let nearestDistance = Number.MAX_VALUE;
   let nearestCity = cities[0].city;
   for (let i = 0; i < cities.length; i++) {
      let distance = getDistanceFromLatLonInM(origin.lat, cities[i].LatLng.lat, origin.lng, cities[i].LatLng.lng);
      console.log(distance);
      if (distance < nearestDistance) {
         nearestDistance = distance;
         nearestCity = cities[i].city;
      }
   }
   console.log(nearestCity);
   return nearestCity;
}

/**
 * This function uses the Haversine formula to try and calculate the distance between two locations based on their latitude and longitude values
 * This algorithm calculates the straight line distance between the two locations.
 * 
 * @param {number} lat1 - The latitude value of the first location
 * @param {number} lat2 - The latitude value of the second location
 * @param {number} lon1 - The longitude value of the first location
 * @param {number} lon2 - The longitude value of the second location
 * @returns {number} - The distance between the locations in meters
 */
export function getDistanceFromLatLonInM(lat1, lat2, lon1, lon2) {
   if (typeof (lat1) !== "number" || typeof (lat2) !== "number" || typeof (lon1) !== "number" || typeof (lon2) !== "number") {
      console.log('The lat/lng must be numbers');
      return null;
   }
   let deg2rad = deg => deg * 0.017453293;
   let a =
      Math.pow(Math.sin(deg2rad(lat2 - lat1) / 2), 2) +
      Math.cos(deg2rad(lat1)) * Math.cos(deg2rad(lat2)) *
      Math.pow(Math.sin(deg2rad(lon2 - lon1) / 2), 2);
   let distance = 12742000 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
   if (distance < 20000000 && distance >= 0) {
      return distance;
   }
   return null;
}

/**
 * This function sorts an array of objects by alphabetical order from A-Z, it sorts by the name key of each object element
 * 
 * @param {Array.<Object>} arrayToBeSorted - The array that will be sorted by alphabetical order
 * @returns {Array.<Object>} - This function returns an array of objects that is sorted by alphabetical order of the name key from A-Z
 */
export function sortAlphaAscending(arrayToBeSorted) {
   //A deep copy of the base array is created
   const temporaryArray = arrayToBeSorted.slice(0);
   for (let i = 0; i < temporaryArray.length; i++) {
      if (typeof (temporaryArray[i].name) !== "string") {
         console.log('sortAlphaAscending must only accept array of objects where the name object is a string')
         return null;
      }
   }
   temporaryArray.sort(function (a, b) {
      var textA = a.name.toUpperCase();
      var textB = b.name.toUpperCase();
      return (textA < textB) ? -1 : (textA > textB) ? 1 : 0;
   });
   console.log(temporaryArray);
   return temporaryArray;
}

/**
 * This function sorts an array of objects by alphabetical order from Z-A, it sorts by the name key of each object element
 *
 * @param {Array.<Object>} arrayToBeSorted - The array that will be sorted by alphabetical order
 * @returns {Array.<Object>} - This function returns an array of objects that is sorted by alphabetical order of the name key from Z-A
 */
export function sortAlphaDescending(arrayToBeSorted) {
   //A deep copy of the base array is created
   const temporaryArray = arrayToBeSorted.slice(0);
   for (let i = 0; i < temporaryArray.length; i++) {
      if (typeof (temporaryArray[i].name) !== "string") {
         console.log('sortAlphaDescending must only accept array of objects where the name object is a string')
         return null;
      }
   }
   temporaryArray.sort(function (a, b) {
      var textA = a.name.toUpperCase();
      var textB = b.name.toUpperCase();
      return (textA < textB) ? 1 : (textA > textB) ? -1 : 0;
   });
   console.log(temporaryArray);
   return temporaryArray;
}

/**
 * This component represents the list
 */
function ListBusiness() {
   const [baseVirtualBusinesses] = useState(businessData.virtualBusinesses.map((business) => {
      return {
         id: business.ID,
         name: business.NAME,
         type: business.TYPE,
         city: business.CITY,
         website: business.WEBSITE,
         description: business.DESCRIPTION,
         contactInfo: business.CONTACT_INFO
      }
   }));
   const [virtualBusinesses, setVirtualBusinesses] = useState([]);
   const [currentPage, setCurrentPage] = useState(1);
   const [dropdownValue, setDropdownValue] = useState("default");
   const [businessesPerPage] = useState(10);
   const [center, setCenter] = useState({
      lat: 0,
      lng: 0
   });

   /**
   * This function calculates the nearest city to the user's location(from the IPAPI) by searching the cities array and finding the city that has the least distance from the user's location
   * 
   * @returns {Array.<Object>} - This functions returns a filtered array of all businesses in the nearest city from the user
   */
   function getNearestCityBusinesses() {
      const nearestCity = getNearestCity(center, allCities);
      console.log(nearestCity);
      //use nearestCity to create a new updated array
      const filteredBusinesses = baseVirtualBusinesses.filter((business) => {
         if (nearestCity === business.city) {
            return business;
         }
         return null;
      });
      console.log(filteredBusinesses);
      //alpha sort below
      const filteredSortedBusinesses = sortAlphaAscending(filteredBusinesses);
      return filteredSortedBusinesses;
   }

   //Initializes the center object when the page loads and initializes the virtualBusinesses array
   useEffect(() => {
      async function fetchCenter() {
         try {
            const cntr = await getUserLocation(IPAPIURL);
            console.log(cntr);
            setCenter(cntr);
         }
         catch (err) {
            setCenter({ lat: 43.5601414, lng: -79.716 });
            alert(err);
         }
      }
      fetchCenter();
      setVirtualBusinesses(baseVirtualBusinesses);
      setCurrentPage(1);
   }, [baseVirtualBusinesses]);

   /**
    * This function executes certain functions depending what the target selectBox value is
    * This function will also reset the pagination page back to the first page
    * @param {EventObject} event 
    */
   const handleChange = (event) => {
      if (event.target.value === "nearestCity") {
         setVirtualBusinesses(getNearestCityBusinesses());
         setCurrentPage(1);
      }
      else if (event.target.value === "default") {
         console.log(baseVirtualBusinesses)
         setVirtualBusinesses(baseVirtualBusinesses);
         setCurrentPage(1);
      }
      else if (event.target.value === "alphaAscending") {
         setVirtualBusinesses(sortAlphaAscending(baseVirtualBusinesses));
         setCurrentPage(1);
      }
      else if (event.target.value === "alphaDescending") {
         setVirtualBusinesses(sortAlphaDescending(baseVirtualBusinesses));
         setCurrentPage(1);
      }
      console.log(event.target.value);
      setDropdownValue(event.target.value);
   }

   //Get Current Businesses(10)
   const indexOfLastBusiness = currentPage * businessesPerPage;
   const indexOfFirstBusiness = indexOfLastBusiness - businessesPerPage;
   const currentBusinesses = virtualBusinesses.slice(indexOfFirstBusiness, indexOfLastBusiness);

   const paginate = pageNumber => setCurrentPage(pageNumber);
   const paginateIncrement = () => setCurrentPage(currentPage + 1);
   const paginateDecrement = () => setCurrentPage(currentPage - 1);

   const pageNumbers = [];

   for (let i = 1; i <= Math.ceil(virtualBusinesses.length / businessesPerPage); i++) {
      pageNumbers.push(i);
   }

   return (
      <div>
         <ScrollToTopOnMount />
         <div className="container">
            <h1 className="title">Virtual Businesses Near Me</h1>

            <h2 className="selectTitle">Sort</h2>
            <div data-testid="sortBy-menu" className="select">
               <select
                  value={dropdownValue}
                  onChange={handleChange}
                  className="select-text"
               >
                  <option value="default">Default</option>
                  <option value="alphaAscending">Name(A-Z)</option>
                  <option value="alphaDescending">Name(Z-A)</option>
                  <option value="nearestCity">Nearest City</option>
               </select>
            </div>
            <Business businesses={currentBusinesses} />
            <div data-testid="pagination" className="pagination">
               <ul>
                  <li>
                     <a
                        data-testid="pagination-previous"
                        className={(currentPage > 1) ? 'button' : 'button inactive'}
                        onClick={paginateDecrement}
                        href="/list-business/#"
                     >
                        &laquo;
                     </a>
                  </li>
                  {pageNumbers.map(number => (
                     <li key={number}>
                        <a
                           className={(currentPage === number) ? 'activePage' : null}
                           onClick={() => paginate(number)} href='/list-business/#'
                        >
                           {number}
                        </a>
                     </li>
                  ))}
                  <li>
                     <a
                        data-testid="pagination-next"
                        className={(currentPage < Math.ceil(virtualBusinesses.length / businessesPerPage)) ? 'button' : 'button inactive'}
                        onClick={paginateIncrement}
                        href="/list-business/#"
                     >
                        &raquo;
                     </a>
                  </li>
               </ul>
            </div>
         </div>
      </div>
   )
}

export default ListBusiness;
