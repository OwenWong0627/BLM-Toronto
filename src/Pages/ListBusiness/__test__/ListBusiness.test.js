import React from 'react';
import ReactDOM from 'react-dom';
import ListBusiness, { getNearestCity, getDistanceFromLatLonInM, sortAlphaAscending, sortAlphaDescending } from '..';

import { BrowserRouter as Router } from 'react-router-dom';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';

//Criteria
//(a) at least 2 working cases,
//(b) at least 1 non-working case,
//(c) any edge/corner cases (like divide by zero, or something that creates infinity, etc.).
afterEach(cleanup);
it('Renders without crashing', () => {
   const div = document.createElement("div");
   ReactDOM.render(<Router><ListBusiness /></Router>, div);
});

it('Renders List Business page with the correct pagination feature(next page/previous page)', () => {
   const { getByTestId } = render(<Router><ListBusiness /></Router>);
   expect(getByTestId('pagination')).toContainHTML("ul");
   expect(getByTestId('pagination')).toContainElement(getByTestId("pagination-previous"));
   expect(getByTestId('pagination')).toContainElement(getByTestId("pagination-next"));
});

it("Renders List Business page's sortBy dropdown with the correct text content", () => {
   const { getByTestId } = render(<Router><ListBusiness /></Router>);
   expect(getByTestId('sortBy-menu')).toHaveTextContent("Default");
   expect(getByTestId('sortBy-menu')).toHaveTextContent("Name(A-Z)");
   expect(getByTestId('sortBy-menu')).toHaveTextContent("Name(Z-A)");
   expect(getByTestId('sortBy-menu')).toHaveTextContent("Nearest City");
})

describe('getNearestCity works as intended', () => {
   const center = { lat: 43.5789, lng: -79.6583 };
   test('getNearestCity calculates the nearest city from an array of city objects', () => {
      let cities = [
         { city: "Mississauga, ON", LatLng: { lat: 43.5769843, lng: -79.7745394 } },
         { city: "Etobicoke, ON", LatLng: { lat: 43.6723023, lng: -79.6928962 } },
         { city: "Brampton, ON", LatLng: { lat: 43.7248485, lng: -79.8996297 } }
      ]
      expect(getNearestCity(center, cities)).toBe("Mississauga, ON");
      expect(getNearestCity(center, cities)).not.toBe("Etobicoke, ON");
   });
   test('getNearestCity calculates the nearest city from cities that are very close to the center', () => {
      let cities = [
         { city: "Close to User City", LatLng: { lat: center.lat + 0.00001, lng: center.lng } },
         { city: "Very Close to User City", LatLng: { lat: center.lat, lng: center.lng - 0.00000000001 } },
         { city: "Very Very Close to User City", LatLng: { lat: center.lat + 0.000000000000000000001, lng: center.lng - 0.00000000000000000001 } }
      ]
      expect(getNearestCity(center, cities)).toBe("Very Very Close to User City");
   });
   test('getNearestCity returns a null value when the lat or lng value is not a number', () => {
      let cities = [{ city: 123456789, LatLng: { lat: "center.lat + 0.00001", lng: "center.lng" } }];
      expect(getNearestCity(center, cities)).toBeNull();
   })
});

describe('getDistanceFromLatLonInM works as intended', () => {
   test('getDistanceFromLatLonInM returns 0 when the two points are identical', () => {
      expect(getDistanceFromLatLonInM(40.31234, 40.31234, -72.420, -72.420)).toBe(0);
   });
   test('getDistanceFromLatLonInM returns a small value in meters when the two points are close to eachother', () => {
      expect(getDistanceFromLatLonInM(40.3123, 40.31234, -72.420, -72.420)).toBeLessThan(1000);
   });
   test('getDistanceFromLatLonInM returns a null value when accepting a non-number parameter', () => {
      expect(getDistanceFromLatLonInM("String", 40.31234, -72.420, -72.420)).toBeNull();
   });
   test('getDistanceFromLatLonInM returns a null value when the distance calculated is longer then the longest distance possible from two points on the Earth', () => {
      expect(getDistanceFromLatLonInM(90, -90, 180, -181)).toBeNull();
   });
});

describe('sortAlphaAscending works as intended', () => {
   test('sortAlphaAscending sorts an array of objects properly from A-Z by the values associated with the name key', () => {
      expect(sortAlphaAscending([{ name: "haha" }, { name: "aaaaa" }, { name: "zzzzz" }])).toMatchObject([{ name: "aaaaa" }, { name: "haha" }, { name: "zzzzz" }]);
      expect(sortAlphaAscending([{ name: "AAAAA" }, { name: "aaaaa" }, { name: "zzzzz" }])).toMatchObject([{ name: "AAAAA" }, { name: "aaaaa" }, { name: "zzzzz" }]);
   });
   test('sortAlphaAscending returns a null value when one of the names of one of the objects is not a string', () => {
      expect(sortAlphaAscending([{ name: 12321323 }, { name: "aaaaa" }, { name: "zzzzz" }])).toBeNull();
   })
});

describe('sortAlphaDescending works as intended', () => {
   test('sortAlphaDescending sorts an array of objects properly from A-Z by the values associated with the name key', () => {
      expect(sortAlphaDescending([{ name: "haha" }, { name: "aaaaa" }, { name: "zzzzz" }])).toMatchObject([{ name: "zzzzz" }, { name: "haha" }, { name: "aaaaa" }]);
      expect(sortAlphaDescending([{ name: "AAAAA" }, { name: "aaaaa" }, { name: "zzzzz" }])).toMatchObject([{ name: "zzzzz" }, { name: "AAAAA" }, { name: "aaaaa" }]);
   });
   test('sortAlphaAscending returns a null value when one of the names of one of the objects is not a string', () => {
      expect(sortAlphaDescending([{ name: 12321323 }, { name: "aaaaa" }, { name: "zzzzz" }])).toBeNull();
   });
});

it('matches snapshot', () => {
   const tree = renderer.create(<Router><ListBusiness /></Router>).toJSON();
   expect(tree).toMatchSnapshot();
});