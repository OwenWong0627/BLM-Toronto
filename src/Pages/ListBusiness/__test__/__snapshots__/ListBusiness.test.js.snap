// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    className="container"
  >
    <h1
      className="title"
    >
      Virtual Businesses Near Me
    </h1>
    <h2
      className="selectTitle"
    >
      Sort
    </h2>
    <div
      className="select"
      data-testid="sortBy-menu"
    >
      <select
        className="select-text"
        onChange={[Function]}
        value="default"
      >
        <option
          value="default"
        >
          Default
        </option>
        <option
          value="alphaAscending"
        >
          Name(A-Z)
        </option>
        <option
          value="alphaDescending"
        >
          Name(Z-A)
        </option>
        <option
          value="nearestCity"
        >
          Nearest City
        </option>
      </select>
    </div>
    <ul
      className="BusinessList"
      data-testid="business-list"
    />
    <div
      className="pagination"
      data-testid="pagination"
    >
      <ul>
        <li>
          <a
            className="button inactive"
            data-testid="pagination-previous"
            href="/list-business/#"
            onClick={[Function]}
          >
            «
          </a>
        </li>
        <li>
          <a
            className="button inactive"
            data-testid="pagination-next"
            href="/list-business/#"
            onClick={[Function]}
          >
            »
          </a>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
